http-server-netty/src/main/java/io/micronaut/http/server/netty/binders/NettyBodyAnnotationBinder.java

```diff
@@ -148,66 +155,73 @@ public List<ConversionError> getConversionErrors() {
+                    buf.release();
                 }
-                intermediate = map;
-            } else if (data.size() == 1) {
-                intermediate = data.get(0);
-                if (intermediate instanceof ByteBufHolder bbh) {
-                    intermediate = bbh.content();
+                processor.complete(data);
+                boolean allFormData = true;
+                for (Object object : data) {
+                    if (!(object instanceof MicronautHttpData<?>)) {
+                        allFormData = false;
+                        break;
+                    }
                 }
-            } else {
-                intermediate = coerceToComposite(data, nhr.getChannelHandlerContext().alloc());
+                Object intermediate;
+                if (allFormData) {
+                    @SuppressWarnings({"unchecked", "rawtypes"})
+                    List<? extends MicronautHttpData<?>> formData = (List) data;
+                    Map<String, Object> map = FormRouteCompleter.toMap(httpServerConfiguration.getDefaultCharset(), formData);
+                    for (MicronautHttpData<?> datum : formData) {
+                        datum.release();
+                    }
+                    intermediate = map;
+                } else if (data.size() == 1) {
+                    intermediate = data.get(0);
+                    if (intermediate instanceof ByteBufHolder bbh) {
+                        intermediate = bbh.content();
+                    }
+                } else {
+                    intermediate = coerceToComposite(data, nhr.getChannelHandlerContext().alloc());
+                }
+                Optional<T> converted =
+                    intermediate instanceof io.netty.util.ReferenceCounted rc ?
+                        NettyConverters.refCountAwareConvert(conversionService, rc, context) :
+                        conversionService.convert(intermediate, context);
+                nhr.setLegacyBody(converted.orElse(null));
+                return converted;
             }
-            Optional<T> converted =
-                intermediate instanceof io.netty.util.ReferenceCounted rc ?
-                    NettyConverters.refCountAwareConvert(conversionService, rc, context) :
-                    conversionService.convert(intermediate, context);
+            if (reader != null) {
+                T result = read(context, reader, nhr.getHeaders(), mediaType, imm.toByteBuffer());
+                nhr.setLegacyBody(result);
+                return Optional.ofNullable(result);
+            }
+            ByteBuf byteBuf = AvailableNettyByteBody.toByteBuf(imm);
+            Optional<T> converted = conversionService.convert(byteBuf, ByteBuf.class, context.getArgument().getType(), context);
+            NettyConverters.postProcess(byteBuf, converted);
             nhr.setLegacyBody(converted.orElse(null));
             return converted;
+        } catch (ContentLengthExceededException t) {
```

Comment:
what is this for?

Reply:
It was in the input stream binder, not sure if it's needed